File: game-model-v2/src/homemade/game/state/impl/CellMap.java
package homemade.game.state.impl;

import homemade.game.fieldstructure.*;
import homemade.game.model.Cell;
import homemade.game.model.CellState;
import homemade.game.model.cellstates.SimpleState;
import homemade.game.state.FieldState;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

abstract public class CellMap extends FieldState {
    private CellState cells[];
    private Link links[];

    private FieldStructure structure;

    private BlockValuePool blockValuePool;

    public CellMap(FieldStructure structure, BlockValuePool blockValuePool) {
        this.structure = structure;
        this.blockValuePool = blockValuePool;

        cells = new CellState[structure.fieldSize];

        CellState emptyState = SimpleState.getSimpleState(Cell.EMPTY);
        for (int i = 0; i < structure.fieldSize; ++i) {
            cells[i] = emptyState;
        }

        int maxLinkCode = structure.numberOfLinks;
        links = new Link[maxLinkCode];

        for (Iterator<LinkCode> iterator = structure.getLinkCodeIterator(); iterator.hasNext(); ) {
            LinkCode link = iterator.next();
            links[link.hashCode()] = new Link();
        }
    }

    @NotNull
    @Override
    public FieldStructure getStructure() {
        return structure;
    }

    @Override
    public CellState getCellState(@NotNull CellCode cellCode) {
        return cells[cellCode.hashCode()];
    }

    @Nullable
    @Override
    public Direction getLinkBetweenCells(@NotNull LinkCode linkCode) {
        return links[linkCode.hashCode()].direction;
    }

    @Override
    public int getChainLength(@NotNull LinkCode linkCode) {
        return links[linkCode.hashCode()].chainLength;
    }

    public void applyCascadeChanges(Map<CellCode, CellState> changes) {
        Set<CellCode> keys = changes.keySet();

        Set<LinkCode> linksToUpdate = new HashSet<>(keys.size() * 4);

        Set<CellState> removedBlocks = new HashSet<>();
        Set<Integer> addedBlockNumbers = new HashSet<>();

        for (CellCode key : keys) {
            CellState newState = changes.get(key);
            CellState oldState = cells[key.hashCode()];

            cells[key.hashCode()] = newState;

            for (Direction direction : Direction.values()) {
                CellCode neighbour = key.neighbour(direction);

                if (neighbour != null) {
                    linksToUpdate.add(structure.getLinkCode(key, neighbour));
                }
            }

            if (oldState.isAliveBlock()) {
                removedBlocks.add(oldState);
            }
            if (newState.isAliveBlock()) {
                addedBlockNumbers.add(newState.value());
            }
        }

        for (CellState removedBlock : removedBlocks) {
            Integer val = removedBlock.value();

            if (!addedBlockNumbers.contains(val)) {
                // so, this is correct, if we process move as single update,
                // and it basically means, that we shouldn't stack removals/additions from different sources
                blockValuePool.freeBlockValue(val);
            }
        }

        for (LinkCode linkCode : linksToUpdate) {
            updateLinkValue(linkCode);
        }
    }

    private void updateLinkValue(LinkCode linkCode) {
        Link link = links[linkCode.hashCode()];

        CellState cellA = cells[linkCode.lower.hashCode()];
        CellState cellB = cells[linkCode.higher.hashCode()];

        Direction oldDirection = link.direction;
        Direction lowerToHigher = linkCode.getLowerToHigherDirection();
        Direction newDirection = linkDirection(cellA.value(), cellB.value(), lowerToHigher);

        link.direction = newDirection;

        Set<Link> alignedLinks = new HashSet<>(structure.getMaxDimension());

        for (Direction direction : EnumSet.of(lowerToHigher, lowerToHigher.getOpposite())) {
            CellCodePair pair = new CellCodePair(linkCode);

            int linksCollected = collectLinks(alignedLinks, pair, direction, newDirection);

            if (pair.isValid() && linksCollected == 1) {
                Set<Link> brokenChainLinks = new HashSet<>(structure.getMaxDimension());
                collectLinks(brokenChainLinks, pair, direction, oldDirection);
                assignLinkLengths(brokenChainLinks, oldDirection);
            }
        }

        assignLinkLengths(alignedLinks, newDirection);
    }

    private Direction linkDirection(int lowerValue, int higherValue, Direction lowerToHigher) {
        Direction direction;

        if (lowerValue == CellState.UNDEFINED_VALUE || higherValue == CellState.UNDEFINED_VALUE ||
                lowerValue == higherValue) {
            direction = null;
        } else if (lowerValue > higherValue) {
            direction = lowerToHigher;
        } else {
            direction = lowerToHigher.getOpposite();
        }

        return direction;
    }

    private int collectLinks(Set<Link> linkSet, CellCodePair pair, Direction moveDirection, Direction chainDirection) {
        int count = 0;
        Link checkLink = links[structure.getLinkCode(pair).hashCode()];

        while (pair.isValid() && checkLink.direction == chainDirection) {
            linkSet.add(checkLink);
            count++;

            pair.move(moveDirection);

            if (pair.isValid()) {
                checkLink = links[structure.getLinkCode(pair).hashCode()];
            }
        }
        return count;
    }

    private void assignLinkLengths(Set<Link> links, Direction direction) {
        int newChainLength = direction == null ? 0 : links.size() + 1;
        for (Link link : links) {
            link.chainLength = newChainLength;
        }
    }

    private static class Link {
        private Direction direction;
        private int chainLength;

        private Link() {
            direction = null;
            chainLength = 0;
        }
    }
}

File: game-model-v2/src/homemade/game/state/impl/CellMap.kt
package homemade.game.state.impl

import homemade.game.fieldstructure.*
import homemade.game.model.Cell
import homemade.game.model.CellState
import homemade.game.model.cellstates.SimpleState
import homemade.game.state.FieldState
import java.util.*

abstract class CellMap(final override val structure: FieldStructure, private val blockValuePool: BlockValuePool) : FieldState() {
    private val cells: Array<CellState>
    private val links: Array<Link?>

    init {
        cells = Array(structure.fieldSize) { SimpleState.get(Cell.EMPTY) }
        val maxLinkCode = structure.numberOfLinks
        links = arrayOfNulls(maxLinkCode)
        val iterator = structure.linkCodeIterator
        while (iterator.hasNext()) {
            val link = iterator.next()
            links[link.hashCode()] = Link()
        }
    }

    override fun getCellState(cellCode: CellCode): CellState {
        return cells[cellCode.hashCode()]
    }

    override fun getLinkBetweenCells(linkCode: LinkCode): Direction? {
        return links[linkCode.hashCode()]!!.direction
    }

    override fun getChainLength(linkCode: LinkCode): Int {
        return links[linkCode.hashCode()]!!.chainLength
    }

    fun applyCascadeChanges(changes: Map<CellCode, CellState?>) {
        val keys = changes.keys
        val linksToUpdate: MutableSet<LinkCode> = HashSet(keys.size * 4)
        val removedBlocks: MutableSet<CellState> = HashSet()
        val addedBlockNumbers: MutableSet<Int> = HashSet()
        for (key in keys) {
            val newState = changes[key]
            val oldState = cells[key.hashCode()]
            cells[key.hashCode()] = newState!!
            for (direction in Direction.values()) {
                val neighbour = key.neighbour(direction)
                if (neighbour != null) {
                    linksToUpdate.add(structure.getLinkCode(key, neighbour))
                }
            }
            if (oldState.isAliveBlock) {
                removedBlocks.add(oldState)
            }
            if (newState.isAliveBlock) {
                addedBlockNumbers.add(newState.value())
            }
        }
        for (removedBlock in removedBlocks) {
            val `val` = removedBlock.value()
            if (!addedBlockNumbers.contains(`val`)) {
                // so, this is correct, if we process move as single update,
                // and it basically means, that we shouldn't stack removals/additions from different sources
                blockValuePool.freeBlockValue(`val`)
            }
        }
        for (linkCode in linksToUpdate) {
            updateLinkValue(linkCode)
        }
    }

    private fun updateLinkValue(linkCode: LinkCode) {
        val link = links[linkCode.hashCode()]
        val cellA = cells[linkCode.lower.hashCode()]
        val cellB = cells[linkCode.higher.hashCode()]
        val oldDirection = link!!.direction
        val lowerToHigher = linkCode.lowerToHigherDirection
        val newDirection = linkDirection(cellA.value(), cellB.value(), lowerToHigher)
        link.direction = newDirection
        val alignedLinks: MutableSet<Link?> = HashSet(structure.maxDimension)
        for (direction in EnumSet.of<Direction?>(lowerToHigher, lowerToHigher.opposite)) {
            val pair = CellCodePair(linkCode)
            val linksCollected = collectLinks(alignedLinks, pair, direction, newDirection)
            if (pair.isValid && linksCollected == 1) {
                val brokenChainLinks: MutableSet<Link?> = HashSet(structure.maxDimension)
                collectLinks(brokenChainLinks, pair, direction, oldDirection)
                assignLinkLengths(brokenChainLinks, oldDirection)
            }
        }
        assignLinkLengths(alignedLinks, newDirection)
    }

    private fun linkDirection(lowerValue: Int, higherValue: Int, lowerToHigher: Direction): Direction? {
        val direction: Direction?
        direction = if (lowerValue == CellState.UNDEFINED_VALUE || higherValue == CellState.UNDEFINED_VALUE || lowerValue == higherValue) {
            null
        } else if (lowerValue > higherValue) {
            lowerToHigher
        } else {
            lowerToHigher.opposite
        }
        return direction
    }

    private fun collectLinks(linkSet: MutableSet<Link?>, pair: CellCodePair, moveDirection: Direction?, chainDirection: Direction?): Int {
        var count = 0
        var checkLink = links[structure.getLinkCode(pair).hashCode()]
        while (pair.isValid && checkLink!!.direction === chainDirection) {
            linkSet.add(checkLink)
            count++
            pair.move(moveDirection!!)
            if (pair.isValid) {
                checkLink = links[structure.getLinkCode(pair).hashCode()]
            }
        }
        return count
    }

    private fun assignLinkLengths(links: Set<Link?>, direction: Direction?) {
        val newChainLength = if (direction == null) 0 else links.size + 1
        for (link in links) {
            link!!.chainLength = newChainLength
        }
    }

    private class Link {
        var direction: Direction? = null
        var chainLength = 0
    }
}
